### PR TITLE
*: add snapshot package

### PR DESCRIPTION
**Pull request by gyuho**
Fix https://github.com/coreos/etcd/issues/8632.
Address https://github.com/coreos/etcd/issues/9109 with tests in `snapshot` package.

### CODE REVIEW

**Comment by xiang90** [comment]
it seems that snapshot should be a sub pkg under etcdctl?

**Comment by Quentin-M** [comment]
That's awesome! Thank you so much!

**Comment by gyuho** [comment]
@xiang90 Added `OutputWALDir` field to `RestoreConfig`

File: snapshot/v3_snapshot.go [hunk 1]
```diff
@@ -0,0 +1,453 @@
...
+	dbPath  string
+	walDir  string
+	snapDir string
+	cl      *membership.RaftCluster
+
+	skipHashCheck bool
+	logger        grpclog.LoggerV2
+}
+
+func (s *v3Snapshotter) Save(ctx context.Context, cli *clientv3.Client, dbPath string) error {
```
Comment (xiang90) [inline snapshot/v3_snapshot.go]: the client should be part of the config i feel, not per save.

File: snapshot/logger.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+import (
+	"log"
+
+	"github.com/coreos/pkg/capnslog"
+	"google.golang.org/grpc/grpclog"
+)
+
+// DefaultLogger defines default snapshot logger based on "capnslog".
+var DefaultLogger grpclog.LoggerV2
```
Comment (xiang90) [inline snapshot/logger.go]: this probably should not belong to snapshot pkg. it is not snapshot related.

File: snapshot/v3_snapshot.go [hunk 3]
```diff
@@ -0,0 +1,456 @@
...
+	"github.com/coreos/etcd/store"
+	"github.com/coreos/etcd/wal"
+	"github.com/coreos/etcd/wal/walpb"
+
+	bolt "github.com/coreos/bbolt"
+	"google.golang.org/grpc/grpclog"
+)
+
+// Snapshotter defines snapshot methods.
+type Snapshotter interface {
```
Comment (xiang90) [inline snapshot/v3_snapshot.go]: rename to Manager? so it looks like snapshot.Manager.Save/Status/Restore

File: snapshot/v3_snapshot.go [hunk 4]
```diff
@@ -0,0 +1,456 @@
...
+
+	// Status returns the snapshot file information.
+	Status(dbPath string) (DBStatus, error)
+
+	// Restore restores a new etcd data directory from given snapshot file.
+	Restore(dbPath string, cfg RestoreConfig) error
+}
+
+// DBStatus is the snapshot file status.
+type DBStatus struct {
```
Comment (xiang90) [inline snapshot/v3_snapshot.go]: Just Status? Hash/Revision is not db (boltdb in our case) specific, but snapshot related.

File: snapshot/v3_snapshot.go [hunk 5]
```diff
@@ -0,0 +1,456 @@
...
+	InitialClusterToken string
+	// PeerURLs is a list of member's peer URLs to advertise to the rest of the cluster.
+	PeerURLs types.URLs
+	// SkipHashCheck is "true" to ignore snapshot integrity hash value
+	// (required if copied from data directory).
+	SkipHashCheck bool
+}
+
+// Config defines snapshot config.
+type Config struct {
```
Comment (xiang90) [inline snapshot/v3_snapshot.go]: we probably should kill this config struct. this is not snapshot related configuration.

File: snapshot/v3_snapshot.go [hunk 6]
```diff
@@ -0,0 +1,456 @@
...
+		return nil
+	}); err != nil {
+		return ds, err
+	}
+
+	ds.Hash = h.Sum32()
+	return ds, nil
+}
+
+type revision struct {
```
Comment (xiang90) [inline snapshot/v3_snapshot.go]: move this to util.go?

File: snapshot/v3_snapshot.go [hunk 7]
```diff
@@ -0,0 +1,456 @@
...
+	mvs.Commit()
+	mvs.Close()
+	be.Close()
+
+	return nil
+}
+
+// initIndex implements ConsistentIndexGetter so the snapshot won't block
+// the new raft instance by waiting for a future raft index.
+type initIndex int
```
Comment (xiang90) [inline snapshot/v3_snapshot.go]: move to util.go?

**Comment by xiang90** [comment]
a few nits. look good to me in general.

**Comment by gyuho** [comment]
@xiang90 PTAL. All fixed.

**Comment by xiang90** [comment]
lgtm once CI gets passed.

**Pull request merged**